File: js/src/feedback.js
/**
 * Feedback operations — the deterministic primitives behind
 * `tux feedback <action>` (SPC sections 42–57).
 *
 * @module feedback
 */
import { CliError, EXIT } from './errors.js';
import { canonicalJson, compactJson } from './canonical.js';
import { nowMs, canonicalTimestamp, newId } from './ids.js';
import { FEEDBACK_TYPES, FEEDBACK_STATUSES, makeFeedback, normalizedText, validateFeedback } from './schema.js';
import {
  loadStore, saveStore, storeExists, ensureSession,
  countIncorporationBatches, saveIncorporationBatch,
} from './store.js';
import { resolveIdentity } from './config.js';

/** Options shared by every action: {cwd, config, identity, format, time}. */
function ctx(opts) {
  return {
    cwd: opts.cwd,
    config: opts.config,
    identity: opts.identity ?? resolveIdentity(opts.config),
    format: opts.format ?? 'text',
    now: opts.now ?? nowMs(),
    timestamp: canonicalTimestamp(opts.now ?? nowMs()),
  };
}

export function opList(opts, filters) {
  const c = ctx(opts);
  const store = loadStore(c.cwd, c.config);
  let items = store.feedback;
  if (filters.status) items = items.filter((f) => f.status === filters.status);
  if (filters.type) items = items.filter((f) => f.feedback.type === filters.type);
  if (filters.mine) items = items.filter((f) => f.author.user_id === c.identity.user_id);
  if (filters.route) items = items.filter((f) => f.location?.route === filters.route);
  if (filters.session) items = items.filter((f) => f.session_id === filters.session);
  if (c.format === 'json') return { stdout: canonicalJson(items) + '\n', exit: EXIT.ok };
  const lines = items.map((f) =>
    [f.id, f.status, f.feedback.type, f.location?.route ?? '-', firstLine(f.feedback.text)].join('\t')
  );
  return { stdout: (lines.length ? lines.join('\n') + '\n' : ''), exit: EXIT.ok };
}

export function opShow(opts, id) {
  const c = ctx(opts);
  const store = loadStore(c.cwd, c.config);
  const item = store.feedback.find((f) => f.id === id);
  if (!item) throw new CliError(EXIT.notFound, `feedback not found: ${id}`);
  return { stdout: canonicalJson(item) + '\n', exit: EXIT.ok };
}

export function opCreate(opts, spec) {
  const c = ctx(opts);
  if (!FEEDBACK_TYPES.includes(spec.type)) {
    throw new CliError(EXIT.usage, `invalid feedback type: ${spec.type} (expected ${FEEDBACK_TYPES.join(', ')})`);
  }
  if (!spec.text || spec.text.trim() === '') throw new CliError(EXIT.usage, '--text is required');
  const store = loadStore(c.cwd, c.config);
  const seq = store.feedback.length + 1;
  const id = newId('fb', store.project_id, spec.session ?? 'default', seq, 'feedback', c.now);
  const item = makeFeedback({
    id,
    project_id: store.project_id,
    session_id: spec.session ?? 'default',
    author: { user_id: c.identity.user_id, display_name: c.identity.display_name },
    origin: spec.origin ?? 'design',
    location: {
      ...(spec.route ? { route: spec.route } : {}),
      ...(spec.page ? { page: spec.page } : {}),
      ...(spec.component ? { component: spec.component } : {}),
      ...(spec.instance ? { component_instance: spec.instance } : {}),
    },
    target: {
      ...(spec.tuxId ? { tux_id: spec.tuxId } : {}),
      ...(spec.testId ? { test_id: spec.testId } : {}),
    },
    ui_state: spec.uiState ?? {},
    type: spec.type,
    text: spec.text,
    created_at: c.timestamp,
  });
  const problems = validateFeedback(item);
  if (problems.length) throw new CliError(EXIT.general, `created item invalid: ${problems.join('; ')}`);
  store.feedback.push(item);
  saveStore(c.cwd, c.config, store);
  if (spec.session) ensureSession(c.cwd, c.config, spec.session, spec.environment ?? 'design', c.timestamp);
  return { stdout: canonicalJson(item) + '\n', exit: EXIT.ok };
}

export function opUpdate(opts, id, patch) {
  const c = ctx(opts);
  const store = loadStore(c.cwd, c.config);
  const item = store.feedback.find((f) => f.id === id);
  if (!item) throw new CliError(EXIT.notFound, `feedback not found: ${id}`);
  if (patch.status !== undefined && !FEEDBACK_STATUSES.includes(patch.status)) {
    throw new CliError(EXIT.usage, `invalid status: ${patch.status} (expected ${FEEDBACK_STATUSES.join(', ')})`);
  }
  if (patch.type !== undefined && !FEEDBACK_TYPES.includes(patch.type)) {
    throw new CliError(EXIT.usage, `invalid feedback type: ${patch.type} (expected ${FEEDBACK_TYPES.join(', ')})`);
  }
  if (patch.text !== undefined) item.feedback.text = patch.text;
  if (patch.type !== undefined) item.feedback.type = patch.type;
  if (patch.status !== undefined) item.status = patch.status;
  if (patch.incorporation !== undefined) item.incorporation = patch.incorporation;
  item.updated_at = c.timestamp;
  saveStore(c.cwd, c.config, store);
  return { stdout: canonicalJson(item) + '\n', exit: EXIT.ok };
}

export function opDelete(opts, id) {
  const c = ctx(opts);
  const store = loadStore(c.cwd, c.config);
  const idx = store.feedback.findIndex((f) => f.id === id);
  if (idx === -1) throw new CliError(EXIT.notFound, `feedback not found: ${id}`);
  store.feedback.splice(idx, 1);
  saveStore(c.cwd, c.config, store);
  if (c.format === 'json') return { stdout: canonicalJson({ deleted: [id] }) + '\n', exit: EXIT.ok };
  return { stdout: `deleted ${id}\n`, exit: EXIT.ok };
}

export function opClear(opts, scope, filters) {
  const c = ctx(opts);
  if (scope === 'all' && !filters.force) {
    throw new CliError(EXIT.usage, '--all requires explicit confirmation (--force)');
  }
  if (!scope) throw new CliError(EXIT.usage, 'specify --mine or --all');
  const store = loadStore(c.cwd, c.config);
  const before = store.feedback.length;
  store.feedback = store.feedback.filter((f) => {
    if (scope === 'mine' && f.author.user_id !== c.identity.user_id) return true;
    if (filters.route && f.location?.route !== filters.route) return true;
    if (filters.session && f.session_id !== filters.session) return true;
    return false;
  });
  const cleared = before - store.feedback.length;
  saveStore(c.cwd, c.config, store);
  if (c.format === 'json') return { stdout: canonicalJson({ cleared }) + '\n', exit: EXIT.ok };
  return { stdout: `cleared ${cleared}\n`, exit: EXIT.ok };
}

export function opExport(opts, format) {
  const c = ctx(opts);
  const store = loadStore(c.cwd, c.config);
  if (format === 'jsonl') {
    const out = store.feedback.map((f) => compactJson(f)).join('\n');
    return { stdout: (out ? out + '\n' : ''), exit: EXIT.ok };
  }
  return { stdout: canonicalJson(store.feedback) + '\n', exit: EXIT.ok };
}

export function storeIsEmpty(opts) {
  return !storeExists(opts.cwd, opts.config);
}

/** Canonical target key for grouping (SPC section 50). */
export function targetKey(item) {
  return [
    item.location?.route ?? '',
    item.location?.page ?? '',
    item.location?.component ?? '',
    item.location?.component_instance ?? '',
    item.target?.tux_id ?? '',
    item.target?.test_id ?? '',
  ].join('|');
}

export function opIncorporate(opts, strategy, filters) {
  const c = ctx(opts);
  const STRATEGIES = ['consolidate', 'requirements', 'tasks', 'direct', 'export-only'];
  if (strategy === undefined) {
    throw new CliError(EXIT.usage, `--strategy is required in non-interactive mode (one of ${STRATEGIES.join(', ')})`);
  }
  if (!STRATEGIES.includes(strategy)) {
    throw new CliError(EXIT.usage, `invalid strategy: ${strategy} (expected ${STRATEGIES.join(', ')})`);
  }
  const store = loadStore(c.cwd, c.config);
  let open = store.feedback.filter((f) => f.status === 'open');
  if (filters.mine) open = open.filter((f) => f.author.user_id === c.identity.user_id);
  if (filters.route) open = open.filter((f) => f.location?.route === filters.route);
  if (filters.session) open = open.filter((f) => f.session_id === filters.session);

  const groups = [];
  const byKey = new Map();
  for (const item of open) {
    const key = targetKey(item);
    if (!byKey.has(key)) {
      const group = { key, items: [] };
      byKey.set(key, group);
      groups.push(group);
    }
    byKey.get(key).items.push(item);
  }

  const reportGroups = [];
  const incorporated = [];
  const conflicts = [];
  for (const group of groups) {
    const items = group.items;
    const kinds = new Set(items.map((i) => i.feedback.type));
    const isConflict = kinds.has('approval') && (kinds.has('change') || kinds.has('issue') || kinds.has('question'));
    if (isConflict) {
      reportGroups.push({ key: group.key, kind: 'conflict', requires_decision: true, feedback_ids: items.map((i) => i.id) });
      for (const i of items) conflicts.push(i.id);
      continue;
    }
    let kind = 'unique';
    if (items.length > 1) {
      const first = normalizedText(items[0].feedback.text);
      kind = items.every((i) => normalizedText(i.feedback.text) === first) ? 'duplicate' : 'unique';
    }
    reportGroups.push({ key: group.key, kind, requires_decision: false, feedback_ids: items.map((i) => i.id) });
    if (strategy !== 'export-only') {
      for (const i of items) {
        i.status = 'incorporated';
        i.incorporation = { strategy, batch_id: null, recorded_at: c.timestamp };
        incorporated.push(i.id);
      }
    }
  }

  const seq = countIncorporationBatches(c.cwd) + 1;
  const batchId = newId('batch', store.project_id, 'default', seq, 'batch', c.now);
  const report = {
    schema_version: '1.0',
    batch_id: batchId,
    strategy,
    created_at: c.timestamp,
    groups: reportGroups,
    incorporated,
    conflicts,
  };
  if (strategy !== 'export-only') {
    for (const item of store.feedback) {
      if (item.incorporation?.batch_id === null && incorporated.includes(item.id)) {
        item.incorporation.batch_id = batchId;
      }
    }
    saveStore(c.cwd, c.config, store);
    saveIncorporationBatch(c.cwd, report);
  }
  return { stdout: canonicalJson(report) + '\n', exit: EXIT.ok };
}

export function opValidate(opts, args) {
  const c = ctx(opts);
  if (args.record !== undefined) {
    if (!['passed', 'failed'].includes(args.result)) {
      throw new CliError(EXIT.usage, `invalid validation result: ${args.result} (expected passed, failed)`);
    }
    const store = loadStore(c.cwd, c.config);
    const item = store.feedback.find((f) => f.id === args.record);
    if (!item) throw new CliError(EXIT.notFound, `feedback not found: ${args.record}`);
    item.validation = { result: args.result, checked_at: c.timestamp, ...(args.note ? { note: args.note } : {}) };
    item.updated_at = c.timestamp;
    saveStore(c.cwd, c.config, store);
    return { stdout: canonicalJson(item) + '\n', exit: EXIT.ok };
  }
  const store = loadStore(c.cwd, c.config);
  const items = store.feedback
    .filter((f) => f.status === 'incorporated')
    .map((f) => ({
      id: f.id,
      location_route: f.location?.route ?? null,
      validation_result: f.validation?.result ?? 'unvalidated',
    }));
  const summary = {
    total: items.length,
    passed: items.filter((i) => i.validation_result === 'passed').length,
    failed: items.filter((i) => i.validation_result === 'failed').length,
    unvalidated: items.filter((i) => i.validation_result === 'unvalidated').length,
  };
  let exit = EXIT.ok;
  if (args.strict && (summary.failed > 0 || summary.unvalidated > 0)) exit = EXIT.general;
  if (c.format === 'json') {
    return { stdout: canonicalJson({ schema_version: '1.0', summary, items }) + '\n', exit };
  }
  const lines = items.map((i) => [i.id, i.validation_result].join('\t'));
  return { stdout: (lines.length ? lines.join('\n') + '\n' : '') + `total ${summary.total}, passed ${summary.passed}, failed ${summary.failed}, unvalidated ${summary.unvalidated}\n`, exit };
}

function firstLine(text) {
  return String(text).split('\n')[0];
}

File: py/tux/feedback.py
"""Feedback operations — the deterministic primitives behind
``tux feedback <action>`` (SPC sections 42–57)."""
from __future__ import annotations

from .canonical import canonical_json, compact_json, sorted_object
from .errors import CliError, Exit
from .ids import canonical_timestamp, new_id, now_ms
from .schema import (FEEDBACK_STATUSES, FEEDBACK_TYPES, SCHEMA_VERSION,
                     make_feedback, normalized_text, validate_feedback)
from .store import (count_incorporation_batches, ensure_session, load_store,
                    save_incorporation_batch, save_store, store_exists)


def _ctx(opts: dict) -> dict:
    now = opts.get("now") or now_ms()
    identity = opts.get("identity") or _resolve_identity(opts["config"])
    return {
        "cwd": opts["cwd"],
        "config": opts["config"],
        "identity": identity,
        "format": opts.get("format") or "text",
        "timestamp": canonical_timestamp(now),
        "_now": now,
    }


def _resolve_identity(config):
    from .config import resolve_identity
    return resolve_identity(config)


def op_list(opts: dict, filters: dict) -> dict:
    c = _ctx(opts)
    store = load_store(c["cwd"], c["config"])
    items = store["feedback"]
    if filters.get("status"):
        items = [f for f in items if f["status"] == filters["status"]]
    if filters.get("type"):
        items = [f for f in items if f["feedback"]["type"] == filters["type"]]
    if filters.get("mine"):
        items = [f for f in items if f["author"]["user_id"] == c["identity"]["user_id"]]
    if filters.get("route"):
        items = [f for f in items if (f.get("location") or {}).get("route") == filters["route"]]
    if filters.get("session"):
        items = [f for f in items if f["session_id"] == filters["session"]]
    if c["format"] == "json":
        return {"stdout": canonical_json(items) + "\n", "exit": Exit.OK}
    lines = [
        "\t".join([f["id"], f["status"], f["feedback"]["type"],
                   (f.get("location") or {}).get("route") or "-",
                   str(f["feedback"]["text"]).split("\n")[0]])
        for f in items
    ]
    return {"stdout": ("\n".join(lines) + "\n") if lines else "", "exit": Exit.OK}


def op_show(opts: dict, feedback_id: str) -> dict:
    c = _ctx(opts)
    store = load_store(c["cwd"], c["config"])
    for item in store["feedback"]:
        if item["id"] == feedback_id:
            return {"stdout": canonical_json(item) + "\n", "exit": Exit.OK}
    raise CliError(Exit.NOT_FOUND, f"feedback not found: {feedback_id}")


def op_create(opts: dict, spec: dict) -> dict:
    c = _ctx(opts)
    ftype = spec.get("type") or "issue"
    if ftype not in FEEDBACK_TYPES:
        raise CliError(Exit.USAGE, f"invalid feedback type: {ftype} (expected {', '.join(FEEDBACK_TYPES)})")
    text = spec.get("text")
    if not text or str(text).strip() == "":
        raise CliError(Exit.USAGE, "--text is required")
    store = load_store(c["cwd"], c["config"])
    session = spec.get("session") or "default"
    seq = len(store["feedback"]) + 1
    feedback_id = new_id("fb", store["project_id"], session, seq, "feedback", c["_now"])
    location = {}
    for key, value in (("route", spec.get("route")), ("page", spec.get("page")),
                       ("component", spec.get("component")),
                       ("component_instance", spec.get("component_instance") or spec.get("instance"))):
        if value:
            location[key] = value
    target = {}
    for key, value in (("tux_id", spec.get("tux_id") or spec.get("tuxId")),
                       ("test_id", spec.get("test_id") or spec.get("testId"))):
        if value:
            target[key] = value
    item = make_feedback({
        "id": feedback_id,
        "project_id": store["project_id"],
        "session_id": session,
        "author": {"user_id": c["identity"]["user_id"], "display_name": c["identity"]["display_name"]},
        "origin": spec.get("origin") or "design",
        "location": location,
        "target": target,
        "ui_state": spec.get("ui_state") or {},
        "type": ftype,
        "text": text,
        "created_at": c["timestamp"],
    })
    problems = validate_feedback(item)
    if problems:
        raise CliError(Exit.GENERAL, "created item invalid: " + "; ".join(problems))
    store["feedback"].append(item)
    save_store(c["cwd"], c["config"], store)
    if spec.get("session"):
        ensure_session(c["cwd"], c["config"], spec["session"], spec.get("environment") or "design", c["timestamp"])
    return {"stdout": canonical_json(item) + "\n", "exit": Exit.OK}


def op_update(opts: dict, feedback_id: str, patch: dict) -> dict:
    c = _ctx(opts)
    store = load_store(c["cwd"], c["config"])
    item = next((f for f in store["feedback"] if f["id"] == feedback_id), None)
    if item is None:
        raise CliError(Exit.NOT_FOUND, f"feedback not found: {feedback_id}")
    if "status" in patch and patch["status"] is not None and patch["status"] not in FEEDBACK_STATUSES:
        raise CliError(Exit.USAGE, f"invalid status: {patch['status']} (expected {', '.join(FEEDBACK_STATUSES)})")
    if "type" in patch and patch["type"] is not None and patch["type"] not in FEEDBACK_TYPES:
        raise CliError(Exit.USAGE, f"invalid feedback type: {patch['type']} (expected {', '.join(FEEDBACK_TYPES)})")
    if patch.get("text") is not None:
        item["feedback"]["text"] = patch["text"]
    if patch.get("type") is not None:
        item["feedback"]["type"] = patch["type"]
    if patch.get("status") is not None:
        item["status"] = patch["status"]
    if patch.get("incorporation") is not None:
        item["incorporation"] = patch["incorporation"]
    item["updated_at"] = c["timestamp"]
    save_store(c["cwd"], c["config"], store)
    return {"stdout": canonical_json(item) + "\n", "exit": Exit.OK}


def op_delete(opts: dict, feedback_id: str) -> dict:
    c = _ctx(opts)
    store = load_store(c["cwd"], c["config"])
    idx = next((i for i, f in enumerate(store["feedback"]) if f["id"] == feedback_id), None)
    if idx is None:
        raise CliError(Exit.NOT_FOUND, f"feedback not found: {feedback_id}")
    store["feedback"].pop(idx)
    save_store(c["cwd"], c["config"], store)
    if c["format"] == "json":
        return {"stdout": canonical_json({"deleted": [feedback_id]}) + "\n", "exit": Exit.OK}
    return {"stdout": f"deleted {feedback_id}\n", "exit": Exit.OK}


def op_clear(opts: dict, scope: str | None, filters: dict) -> dict:
    c = _ctx(opts)
    if scope == "all" and not filters.get("force"):
        raise CliError(Exit.USAGE, "--all requires explicit confirmation (--force)")
    if not scope:
        raise CliError(Exit.USAGE, "specify --mine or --all")
    store = load_store(c["cwd"], c["config"])
    before = len(store["feedback"])

    def keep(f):
        if scope == "mine" and f["author"]["user_id"] != c["identity"]["user_id"]:
            return True
        if filters.get("route") and (f.get("location") or {}).get("route") != filters["route"]:
            return True
        if filters.get("session") and f["session_id"] != filters["session"]:
            return True
        return False

    store["feedback"] = [f for f in store["feedback"] if keep(f)]
    cleared = before - len(store["feedback"])
    save_store(c["cwd"], c["config"], store)
    if c["format"] == "json":
        return {"stdout": canonical_json({"cleared": cleared}) + "\n", "exit": Exit.OK}
    return {"stdout": f"cleared {cleared}\n", "exit": Exit.OK}


def op_export(opts: dict, fmt: str) -> dict:
    c = _ctx(opts)
    store = load_store(c["cwd"], c["config"])
    if fmt == "jsonl":
        lines = [compact_json(f) for f in store["feedback"]]
        return {"stdout": ("\n".join(lines) + "\n") if lines else "", "exit": Exit.OK}
    return {"stdout": canonical_json(store["feedback"]) + "\n", "exit": Exit.OK}


def target_key(item) -> str:
    loc = item.get("location") or {}
    target = item.get("target") or {}
    return "|".join([
        loc.get("route") or "", loc.get("page") or "",
        loc.get("component") or "", loc.get("component_instance") or "",
        target.get("tux_id") or "", target.get("test_id") or "",
    ])


def op_incorporate(opts: dict, strategy: str | None, filters: dict) -> dict:
    c = _ctx(opts)
    strategies = ["consolidate", "requirements", "tasks", "direct", "export-only"]
    if strategy is None:
        raise CliError(Exit.USAGE, f"--strategy is required in non-interactive mode (one of {', '.join(strategies)})")
    if strategy not in strategies:
        raise CliError(Exit.USAGE, f"invalid strategy: {strategy} (expected {', '.join(strategies)})")
    store = load_store(c["cwd"], c["config"])
    open_items = [f for f in store["feedback"] if f["status"] == "open"]
    if filters.get("mine"):
        open_items = [f for f in open_items if f["author"]["user_id"] == c["identity"]["user_id"]]
    if filters.get("route"):
        open_items = [f for f in open_items if (f.get("location") or {}).get("route") == filters["route"]]
    if filters.get("session"):
        open_items = [f for f in open_items if f["session_id"] == filters["session"]]

    groups: list[dict] = []
    by_key: dict[str, dict] = {}
    for item in open_items:
        key = target_key(item)
        if key not in by_key:
            group = {"key": key, "items": []}
            by_key[key] = group
            groups.append(group)
        by_key[key]["items"].append(item)

    report_groups = []
    incorporated = []
    conflicts = []
    for group in groups:
        items = group["items"]
        kinds = {i["feedback"]["type"] for i in items}
        is_conflict = "approval" in kinds and bool(kinds & {"change", "issue", "question"})
        if is_conflict:
            report_groups.append({"key": group["key"], "kind": "conflict",
                                  "requires_decision": True, "feedback_ids": [i["id"] for i in items]})
            conflicts.extend(i["id"] for i in items)
            continue
        kind = "unique"
        if len(items) > 1:
            first = normalized_text(items[0]["feedback"]["text"])
            kind = "duplicate" if all(normalized_text(i["feedback"]["text"]) == first for i in items) else "unique"
        report_groups.append({"key": group["key"], "kind": kind,
                              "requires_decision": False, "feedback_ids": [i["id"] for i in items]})
        if strategy != "export-only":
            for i in items:
                i["status"] = "incorporated"
                i["incorporation"] = {"strategy": strategy, "batch_id": None, "recorded_at": c["timestamp"]}
                incorporated.append(i["id"])

    seq = count_incorporation_batches(c["cwd"]) + 1
    batch_id = new_id("batch", store["project_id"], "default", seq, "batch", c["_now"])
    report = {
        "schema_version": SCHEMA_VERSION,
        "batch_id": batch_id,
        "strategy": strategy,
        "created_at": c["timestamp"],
        "groups": report_groups,
        "incorporated": incorporated,
        "conflicts": conflicts,
    }
    if strategy != "export-only":
        for item in store["feedback"]:
            inc = item.get("incorporation")
            if inc is not None and inc.get("batch_id") is None and item["id"] in incorporated:
                inc["batch_id"] = batch_id
        save_store(c["cwd"], c["config"], store)
        save_incorporation_batch(c["cwd"], report)
    return {"stdout": canonical_json(report) + "\n", "exit": Exit.OK}


def op_validate(opts: dict, args: dict) -> dict:
    c = _ctx(opts)
    if args.get("record") is not None:
        result = args.get("result")
        if result not in ("passed", "failed"):
            raise CliError(Exit.USAGE, f"invalid validation result: {result} (expected passed, failed)")
        store = load_store(c["cwd"], c["config"])
        item = next((f for f in store["feedback"] if f["id"] == args["record"]), None)
        if item is None:
            raise CliError(Exit.NOT_FOUND, f"feedback not found: {args['record']}")
        validation = {"result": result, "checked_at": c["timestamp"]}
        if args.get("note"):
            validation["note"] = args["note"]
        item["validation"] = validation
        item["updated_at"] = c["timestamp"]
        save_store(c["cwd"], c["config"], store)
        return {"stdout": canonical_json(item) + "\n", "exit": Exit.OK}
    store = load_store(c["cwd"], c["config"])
    items = [
        {"id": f["id"],
         "location_route": (f.get("location") or {}).get("route"),
         "validation_result": (f.get("validation") or {}).get("result") or "unvalidated"}
        for f in store["feedback"] if f["status"] == "incorporated"
    ]
    summary = {
        "total": len(items),
        "passed": len([i for i in items if i["validation_result"] == "passed"]),
        "failed": len([i for i in items if i["validation_result"] == "failed"]),
        "unvalidated": len([i for i in items if i["validation_result"] == "unvalidated"]),
    }
    exit_code = Exit.OK
    if args.get("strict") and (summary["failed"] > 0 or summary["unvalidated"] > 0):
        exit_code = Exit.GENERAL
    if c["format"] == "json":
        payload = {"schema_version": SCHEMA_VERSION, "summary": summary, "items": items}
        return {"stdout": canonical_json(payload) + "\n", "exit": exit_code}
    lines = [f"{i['id']}\t{i['validation_result']}" for i in items]
    text = ("\n".join(lines) + "\n") if lines else ""
    text += (f"total {summary['total']}, passed {summary['passed']}, "
             f"failed {summary['failed']}, unvalidated {summary['unvalidated']}\n")
    return {"stdout": text, "exit": exit_code}
